### PR TITLE
Fix issues while CASTing string literal to datetime, datetime2 and datetimeoffset

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

This PR addresses the behavioural differences in CASTing of string literals to datetime, datetime2 and datetimeoffset. The following changes have been made to fix the issue : 
1. Currently, we use native PG function named `ParseDateTime` for parsing the input string literals. This function does not handle some cases such as empty spaces between the fields. We have created a function named `cleanup_input_str` which will modify the input string to remove all the unnecessary spaces and then call the inbuilt `ParseDateTime` function. 
2. `ParseDatetime` splits the input string to different fields which then will be decoded in native function named`DecodeDateTime`.  But `DecodeDatetime` will not handle some cases such as accepting `:` as a separator for seconds and milliseconds. These additional cases were handled in a different function named `tsql_decode_datetime_fields` .
3. Also, if there is a text month we check the REGEX of the input str so that it matches the T-SQL behaviour

### Issues Resolved

BABEL-4328
Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -** YES

* **Arbitrary inputs -** YES

* **Negative test cases -** YES

* **Minor version upgrade tests -** YES

* **Major version upgrade tests -** YES

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_common/src/datetime.h [hunk 1]
```diff
@@ -43,4 +43,42 @@ extern bool int32_multiply_add(int32 val, int32 multiplier, int32 *sum);
 
 extern Datum datetime_in_str(char *str, Node *escontext);
 
+#define DATE_TIME 1
+#define DATE_TIME_2 2
+#define DATE_TIME_OFFSET 3
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.h]: Can be an enum

File: contrib/babelfishpg_common/src/datetime.h [hunk 2]
```diff
@@ -43,4 +43,42 @@ extern bool int32_multiply_add(int32 val, int32 multiplier, int32 *sum);
...
+    "[0-9]{2}\\s*[a-zA-Z]{3,5}", // yy mon
+    "[0-9]{4}\\s*[-]\\s*[a-zA-Z]{3,5}\\s*[-]\\s*[0-9]{1,2}", // yyyy-mon-dd
+    "[0-9]{4}\\s*[/]\\s*[a-zA-Z]{3,5}\\s*[/]\\s*[0-9]{1,2}", // yyyy/mon/dd
+    "[0-9]{4}\\s*[.]\\s*[a-zA-Z]{3,5}\\s*[.]\\s*[0-9]{1,2}", // yyyy.mon.dd
+    "[0-9]{1,2}\\s*[-]\\s*[a-zA-Z]{3,5}\\s*[-]\\s*[0-9]{4}", // dd-mon-[yy]yy
+    "[0-9]{1,2}\\s*[/]\\s*[a-zA-Z]{3,5}\\s*[/]\\s*[0-9]{4}", // dd/mon/[yy]yy
+    "[0-9]{1,2}\\s*[.]\\s*[a-zA-Z]{3,5}\\s*[.]\\s*[0-9]{4}" // dd.mon.[yy]yy
+};
+
+static const int num_date_regexes = sizeof(regex_date_set) / sizeof(regex_date_set[0]);
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.h]: 1. Declare it as a MACRO instead.
2. Above array be named as `date_regexes` and this macro can be named as `NUM_DATE_REGEXES`, also can use `lengthof` optionally:
```
#define NUM_DATE_REGEXES lengthof(date_regexes)
```

File: contrib/babelfishpg_common/src/datetime.h [hunk 3]
```diff
@@ -43,4 +43,42 @@ extern bool int32_multiply_add(int32 val, int32 multiplier, int32 *sum);
...
+
+static const char *regex_time_set[] = {
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.h]: Can be named as `time_regexes`

File: contrib/babelfishpg_common/src/datetime.h [hunk 4]
```diff
@@ -43,4 +43,42 @@ extern bool int32_multiply_add(int32 val, int32 multiplier, int32 *sum);
...
+    "[0-9]{1,2}:[0-9]{1,2}\\s*([AP]M)?", // hh:mm
+    "[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}\\s*([AP]M)?", // hh:mm:ss
+    "[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}[.][0-9]{1,9}\\s*([AP]M)?", // hh:mm:ss.fffff
+    "[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}:[0-9]{1,9}\\s*([AP]M)?", // hh:mm:ss:fffff
+    "[0-9]{1,2}\\s*([AP]M)" // hh AM/PM
+};
+
+static const int num_time_regexes = sizeof(regex_time_set) / sizeof(regex_time_set[0]);
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.h]: Can be a macro `NUM_TIME_REGEXES`

File: contrib/babelfishpg_common/src/datetime.c [hunk 5]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
 
 #define DTK_NANO 32
 
+static const char *regex_time_offset = "(((\\+|\\-)[0-9]{1,2}:[0-9]{1,2})|Z)?";
+
+int
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c]: 1. Let's add a comment here to explain what this function does.
2. Looks like this function returns 0/1 so can change the return type to `bool`

File: contrib/babelfishpg_common/src/datetime.c [hunk 6]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
 
 #define DTK_NANO 32
 
+static const char *regex_time_offset = "(((\\+|\\-)[0-9]{1,2}:[0-9]{1,2})|Z)?";
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c]: This need not be declared globally if this being used only in one function, make it local instead.

File: contrib/babelfishpg_common/src/datetime.c [hunk 7]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
...
+
+            		if (status == 0)
+                		return 1;
+        	}
+    	}
+
+	return 0;
+}
+
+char*
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c:174]: Let's add a comment above to explain what this function does.

File: contrib/babelfishpg_common/src/datetime.c [hunk 8]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
...
+	int num_colons = 0;
+
+	while (str[i] != '\0')
+	{
+		/*
+		 * Remove leading spaces
+		 */
+		while (str[i] != '\0' && isspace(str[i]))
+		{
+			if (i != 0)
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c:191]: Question: Why not counting first space as `contains_extra_spaces`?

File: contrib/babelfishpg_common/src/datetime.c [hunk 9]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
...
+
+	return 0;
+}
+
+char*
+clean_input_str(char *str, bool *contains_extra_spaces, int context)
+{
+	char *result = (char *) palloc(MAXDATELEN);
+	int i = 0, j = 0;
+	int last_non_space = -1;
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c:179]: What does `last_non_space` mean?

File: contrib/babelfishpg_common/src/datetime.c [hunk 10]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
...
+
+		/*
+		 * Modify DATE delimiters to '.'
+		 */
+		if (context == DATE_TIME && (str[i] == '/' || str[i] == '-' || str[i] == '.'))
+		{
+			result[j] = '.';
+			j++;
+		}
+		else if (isalpha(str[i]))
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c:226]: Let's have a comment before each `if-else` to explain what are we doing for better readability.

File: contrib/babelfishpg_common/src/datetime.c [hunk 11]
```diff
@@ -63,6 +64,344 @@ void		CheckDatetimePrecision(fsec_t fsec);
...
+		i++;
+	}
+
+	result[j] = '\0';
+
+	return result;
+	
+}
+
+static int
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c]: Looks like the function only returns 0/1 so return type better be `bool`.

File: contrib/babelfishpg_common/src/datetime.c [hunk 12]
```diff
@@ -92,20 +434,56 @@ datetime_in_str(char *str, Node *escontext)
...
+		if (!is_year_set)
+			tm->tm_year = 1900;
+		if (!tm->tm_mon)
+			tm->tm_mon = 1;
+		if (is_year_set || !tm->tm_mday)
+			tm->tm_mday = 1;
+		dterr = 0;
+	}
+	else if(dterr == 1)
 	{
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime.c:592]: I think we can keep the initial `if (dterr == 1)` condition and add additional condition for `is_year_set` inside it.

File: contrib/babelfishpg_common/src/datetime2.c [hunk 13]
```diff
@@ -36,6 +37,189 @@ static void AdjustDatetime2ForTypmod(Timestamp *time, int32 typmod);
 static Datum datetime2_in_str(char *str, int32 typmod, Node *escontext);
 void		CheckDatetime2Range(const Timestamp time, Node *escontext);
 
+int
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetime2.c]: Same as above, return type can be `bool`

File: contrib/babelfishpg_common/src/datetimeoffset.c [hunk 14]
```diff
@@ -68,8 +70,8 @@ PG_FUNCTION_INFO_V1(get_datetimeoffset_tzoffset_internal);
 PG_FUNCTION_INFO_V1(dateadd_datetimeoffset);
 
 #define DTK_NANO 32
-
-
+ 
+
```
Comment (rishabhtanwar29) [inline contrib/babelfishpg_common/src/datetimeoffset.c]: Unnecessary change.

File: test/JDBC/upgrade/latest/schedule [hunk 15]
```diff
@@ -529,4 +529,7 @@ babel_test_int8_numeric_oper
 babel_test_int2_numeric_oper
 login_token-dep
 test_like_for_AI
-BABEL-4869
\ No newline at end of file
+BABEL-4869
+babel_4328_datetime
+babel_4328_datetime2
+babel_4328_datetimeoffset
```
Comment (rishabhtanwar29) [inline test/JDBC/upgrade/latest/schedule]: We also need to add these tests to `15_7/schedule` and `14_12/schedule` files when we cherry-pick this to previous versions. So, let's raise the PRs for 3.x and 2.x and then we will merge them from 2.x to 4.x in that order.

File: test/JDBC/expected/babel_4328_datetime-before-16_3-vu-prepare.out [hunk 16]
```diff
@@ -0,0 +1,74 @@
+create view babel_4328_datetime_v1 as select cast('2020' as datetime)
```
Comment (rishabhtanwar29) [inline test/JDBC/expected/babel_4328_datetime-before-16_3-vu-prepare.out:1]: @rohit01010 is working on the fixing the `convert` for this so please cross check all the test cases to make sure both of you use all the test cases combined to enhance the test coverage. Refer this PR https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2571

**Review by rishabhtanwar29 — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]
